### PR TITLE
Fixed #32236 -- Made LoginRequiredMixin/PermissionRequiredMixin subclass UserPassesTestMixin.

### PR DESCRIPTION
**Pull request by jerinpetergeorge**
Ticket: https://code.djangoproject.com/ticket/32236

### CODE REVIEW

File: django/contrib/auth/mixins.py [hunk 1]
```diff
@@ -63,15 +63,52 @@ def handle_no_permission(self):
...
+    """
+    Deny a request with a permission error if the test_func() method returns
+    False.
+    """
+
+    def test_func(self):
+        """
+        Write custom logic of your test and return either True or False
+        """
+        return True
```
Comment (felixxm) [inline django/contrib/auth/mixins.py:76]: Why we change the logic here? This patch shouldn't change anything in the `UserPassesTestMixin` implementation.

File: django/contrib/auth/mixins.py [hunk 2]
```diff
@@ -63,15 +63,52 @@ def handle_no_permission(self):
...
+        This function likely to be deprecated in future,
+        use `get_test_functions()` method instead
+        """
+        return self.test_func
+
+    def get_test_functions(self):
+        if hasattr(self, "get_test_func"):
+            # DeprecationWarning: `get_test_func` likely to be deprecated in future
+            return [self.get_test_func()]
+        return [self.test_func]
```
Comment (felixxm) [inline django/contrib/auth/mixins.py:91]: Do we need a new method? and why we want to deprecate the current one?

**Review by felixxm — COMMENTED** [review]
@jerinpetergeorge Thanks for this patch, however it shouldn't change anything in the `UserPassesTestMixin` implementation. I'm not sure why you want to add `get_test_functions()` and deprecate `get_test_func()`. This ticket is only about subclassing `UserPassesTestMixin` by `LoginRequiredMixin/PermissionRequiredMixin` which is achievable without any changes in the `UserPassesTestMixin`.

Extra tests and docs change are not required.

**Comment by jerinpetergeorge** [comment]
Thanks for the review @felixxm 

Initially, I have updated the `LoginRequiredMixin` class and `PermissionRequiredMixin` by subclassing them from `UserPassesTestMixin`, as the ticket intended. But, that made the `AccessMixinTests.test_stacked_mixins_not_logged_in()` test cases failed :astonished: 
_________________
Suppose we have a situation like below,
```python
class LoginRequiredMixin(UserPassesTestMixin):
    """Verify that the current user is authenticated."""

    def is_authenticated(self):
        return self.request.user.is_authenticated

    def get_test_func(self):
        return self.is_authenticated


class PermissionRequiredMixin(UserPassesTestMixin):
    """Verify that the current user has all specified permissions."""
    permission_required = None

    def get_permission_required(self):
        """
        Override this method to override the permission_required attribute.
        Must return an iterable.
        """
        if self.permission_required is None:
            raise ImproperlyConfigured(
                '{0} is missing the permission_required attribute. Define {0}.permission_required, or override '
                '{0}.get_permission_required().'.format(self.__class__.__name__)
            )
        if isinstance(self.permission_required, str):
            perms = (self.permission_required,)
        else:
            perms = self.permission_required
        return perms

    def has_permission(self):
        """
        Override this method to customize the way permissions are checked.
        """
        perms = self.get_permission_required()
        return self.request.user.has_perms(perms)

    def get_test_func(self):
        return self.has_permission


from django.views.generic import View


class IsSuperUser(UserPassesTestMixin):
    def test_func(self):
        return self.request.user.is_superuser


class StackedView(
    LoginRequiredMixin,
    PermissionRequiredMixin,
    IsSuperUser,
    View
):
    permission_required = [
        'auth_tests.add_customuser',
        'auth_tests.change_customuser'
    ]
    raise_exception = True
```

If we consider the **`StackedView`**, only the `is_authenticated(...)` method of  `LoginRequiredMixin` is gets executed because of how the mro works in Python (correct me if I am wrong here). 

So, I thought of calling the `super()` method of `get_test_func()` is the appropriate way to tackle the situation. Again, that resulted in another issue. The `get_test_func()` supposed to return only a single *function*. At last, I reached this solution.

**Comment by jerinpetergeorge** [comment]
This issue has already been mentioned in [our doc](https://docs.djangoproject.com/en/3.1/topics/auth/default/) under "Stacking UserPassesTestMixin" section. So, I am afraid that the *"proposed solution"* will result in a situation where we will not able to use the `LoginRequiredMixin` and `PermissionRequiredMixin` in a *stacked* fashion.

**Comment by felixxm** [comment]
@jerinpetergeorge Thanks for details :+1: I think we should _wontfix_ this ticket, isn't worth the complexity.

Closing per ticket.

**Review by sardar7110 — COMMENTED** [review]
#-

**Pull request closed without merging**